Improve tracker entries handling

glassez opened the pull request:
Closes #19443.

glassez commented:
Created on top of #19493 so it should be merged first.

stalkerok commented:
There must be something wrong with the values, they are displayed in this format.
![2023-08-22_194205](https://github.com/qbittorrent/qBittorrent/assets/63958081/dcd15095-2572-4226-b828-90bee7cd8bbb)

glassez commented:
> There must be something wrong with the values, they are displayed in this format.

It is time of the next (min) announce. I dropped the date for brevity, because it is usually (IIRC) either the same day or the next, which is easily understandable, knowing the current time.

stalkerok commented:
If it's just time, then it obviously doesn't match what's on the screen (8 pm).

glassez commented:
> I dropped the date for brevity, because it is usually (IIRC) either the same day or the next, which is easily understandable, knowing the current time.

Forget it. Switched to locale dependent (short) datetime format.

glassez commented:
> If it's just time, then it obviously doesn't match what's on the screen (8 pm).

What is your time zone?
Could you try it now?

stalkerok commented:
> What is your time zone?
> Could you try it now?

GMT+3 (Brest)
I'm going to try now.

stalkerok commented:
https://user-images.githubusercontent.com/63958081/262443213-bb6da1cc-3dd0-4a0b-b70d-5e2db172d614.png

stalkerok commented:
https://user-images.githubusercontent.com/63958081/262443554-a1f851b7-21b5-40cb-9ec3-40069ed9b187.png

glassez commented:
@stalkerok 
One more time, please?
I cannot test in under Windows currently. It works correctly under Linux all the way.

stalkerok commented:
Great! The only negative is that when collapsed, the time of the next announce is displayed from an unreachable endpoint.
![2023-08-22_220738](https://github.com/qbittorrent/qBittorrent/assets/63958081/53a4af68-e5b6-4799-a214-ec293d6fb3a6)
![2023-08-22_220721](https://github.com/qbittorrent/qBittorrent/assets/63958081/033e3f43-a6d5-4ed5-a461-34bdcef4eb45)

glassez commented:
> The only negative is that when collapsed, the time of the next announce is displayed from an unreachable endpoint.

It shows the date of the next announce attempt. What would be more expected to see it instead? The date of the next announce through one of the currently working endpoints?

glassez commented:
> What would be more expected to see it instead? The date of the next announce through one of the currently working endpoints?

Could you test it now?

glassez commented:
BTW, wouldn't be a good idea to completely exclude unreachable endpoints from being processed/displayed?
@kasper93? @qbittorrent/bug-handlers?

stalkerok commented:
> Could you test it now?

Already better
![2023-08-23_092940](https://github.com/qbittorrent/qBittorrent/assets/63958081/959a48b4-a81e-4d7d-88ac-2222eb30313e)
.
![2023-08-23_092953](https://github.com/qbittorrent/qBittorrent/assets/63958081/a2010616-e9d6-4327-a76c-73ece9f7b39d)
Only where the tracker responded unregistered does the time show from the unreachable endpoint.
If the tracker is in working condition, then it is displayed normally.
![2023-08-23_092737](https://github.com/qbittorrent/qBittorrent/assets/63958081/97e55dec-d3a3-4d0a-9daa-eb00e02c6d77)
.
![2023-08-23_092751](https://github.com/qbittorrent/qBittorrent/assets/63958081/282cc518-a071-433a-bb76-6e292cea7cd5)
This is already good enough, probably the best thing would be to remove the display of unreachable endpoints.

glassez commented:
> probably the best thing would be to remove the display of unreachable endpoints.

OK, will provide separate PR for it.

glassez commented:
> The reasoning behind that is the fact that announces are most likely to happen on the same day, unless we're looping over to next day. So date is a less significant metric here.

I could display only time component (as I was done initially).

> Also displaying only the time remaining, negates the requirement for manual calculation.

What "manual calculation" do you talk about? Do you need to calculate something when you see the time of the next announcement?

In fact, it's easier for me to show it as a specific date/time, because `libtorrent` provides it in this form. In addition, if you display the remaining time, it will require updating it every small period of time.

Chocobo1 commented on src/gui/properties/trackerlistwidget.cpp:
```diff
@@ -476,13 +493,17 @@ void TrackerListWidget::loadTrackers()
             item->setText(COL_LEECHES, item->data(COL_LEECHES, Qt::UserRole).toString());
             item->setText(COL_TIMES_DOWNLOADED, item->data(COL_TIMES_DOWNLOADED, Qt::UserRole).toString());
             item->setText(COL_MSG, item->data(COL_MSG, Qt::UserRole).toString());
+            item->setText(COL_NEXT_ANNOUNCE, item->data(COL_NEXT_ANNOUNCE, Qt::UserRole).toString());
+            item->setText(COL_MIN_ANNOUNCE, item->data(COL_MIN_ANNOUNCE, Qt::UserRole).toString());
```
Have you considered using the format of `Utils::Misc::userFriendlyDuration()` instead?

glassez replied:
`Utils::Misc::userFriendlyDuration()` is for representing duration, not datetime (timepoint).

Chocobo1 replied:
Just to clarify. I suppose user is more interested in when the next announce is going to happen and showing 5mins (later) is intuitive than hh:mm:ss format, at least for myself.

Chocobo1 commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -78,6 +78,16 @@ namespace
+    {
+        using namespace std::chrono;
+
+        const auto ltNow = time_point_cast<lt::seconds32>(lt::clock_type::now());
+        const auto secsSinceNow = duration_cast<seconds>(timePoint - ltNow).count();
+        const auto qNow = QDateTime::currentDateTime();
```
I would try to avoid calling both `QDateTime::currentDateTime()` and `lt::clock_type::now()`. There are system calls within and should be minimized.
Perhaps `QDateTime::fromStdTimePoint()` and its overloads can help here?

glassez replied:
> Perhaps `QDateTime::fromStdTimePoint()` and its overloads can help here?

Unfortunately it is unavailable with the compiler I use currently on Linux.
I'll try to figure out if there is an acceptable opportunity to update it.

glassez replied:
Unfortunately I can't use `QDateTime::fromStdTimePoint()` with MSVC2022 too.

Chocobo1 replied:
>Unfortunately I can't use QDateTime::fromStdTimePoint() with MSVC2022 too.

Probably because the clocks isn't compatible.

Please try this:
```c++
#include "libtorrent/aux_/time.hpp"

QDateTime fromLTTimePoint32(const lt::time_point32 &timePoint)
{
    const std::time_t seconds = lt::aux::to_time_t(timePoint);
    return QDateTime::fromSecsSinceEpoch(seconds);
}
```

glassez replied:
`lt::aux::to_time_t()` isn't exported in regular builds. Anyway there are two `now()` there too so what's the point?

stalkerok commented:
Such a model would be really better, wouldn't have to look at the current time and calculate on your own how much time remains until the next announcement.

glassez commented:
> Such a model would be really better, wouldn't have to look at the current time and calculate on your own how much time remains until the next announcement.

Since you all like it better, I'll change it.

johnyb0y commented:
How about using the general WebUI update interval which is already found in the settings?

glassez commented:
> > Such a model would be really better, wouldn't have to look at the current time and calculate on your own how much time remains until the next announcement.
> 
> Since you all like it better, I'll change it.

Done.

glassez commented:
> It remains to do as with the messages of the tracker, so that the countdown would show from it in a collapsed form.

Sorry, I can't understand it. Could you rephrase it?

stalkerok commented:
I mean, in a collapsed form, the timer of the next announcement shows the time from an unreachable checkpoint on those non-working trackers that return an error.

glassez commented:
> in a collapsed form, the timer of the next announcement shows the time from an unreachable endpoint on those non-working trackers that return an error.

Would it make more sense if it shows the time of the endpoint that has the tracker error? Does "next announce" even make sense for such "unregistered" torrents?

Chocobo1 commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -78,6 +78,16 @@ namespace
 
+    QDateTime fromLTTimePoint32(const lt::time_point32 &timePoint)
+    {
+        using namespace std::chrono;
+
+        const auto ltNow = lt::clock_type::now();
+        const auto secsSinceNow = duration_cast<seconds>(timePoint - ltNow + lt::milliseconds(500)).count();
+        const auto qNow = QDateTime::currentDateTime();
+        return qNow.addSecs(secsSinceNow);
```
Still, I think you should stick to the same namespace even if the underlying is the same.
```suggestion
        const auto ltNow = lt::clock_type::now();
        const auto qNow = QDateTime::currentDateTime();  // moved to minimize time difference
        const auto secsSinceNow = lt::duration_cast<lt::seconds>(timePoint - ltNow + lt::milliseconds(500)).count();
        return qNow.addSecs(secsSinceNow);
```

glassez replied:
> Still, I think you should stick to the same namespace even if the underlying is the same.

Ok. Done.

stalkerok commented:
> Would it make more sense if it shows the time of the endpoint that has the tracker error?

Yes

> Does "next announce" even make sense for such "unregistered" torrents?

It makes sense, the torrent can be registered on the tracker.

glassez commented:
> > Does "next announce" even make sense for such "unregistered" torrents?
> 
> It makes sense, the torrent can be registered on the tracker.

Strange. I assumed that the "unregistered" state usually means the end of the torrent's life.

stalkerok commented:
You can add a tracker to the torrent before it is registered. The same thing, for example, with the error "banned client", it can be entered and removed from the "white list".

stalkerok commented:
Wow, thanks, this is already very good!
https://user-images.githubusercontent.com/63958081/264990756-a10fb73f-4505-4241-90eb-afabdaa911bf.png
Of the entire list of trackers tested, only one shows a message from an unreachable endpoint instead of an error message.
https://user-images.githubusercontent.com/63958081/264993908-c5388796-64f4-40c8-a6f7-53faa8d35d08.png

glassez commented:
>Of the entire list of trackers tested, only one shows a message from an unreachable endpoint instead of an error message.

Could you test it now?

stalkerok commented:
> Could you test it now?

Now everything is fine.

glassez commented:
> When a torrent is in a paused state, it continues to display the next announce time, and when it reaches zero, it shows the infinity symbol.

It seems to be a kind of "memory effect". When the torrent stops, no more tracker related events occur, so we have the last values received. Moreover, `libtorrent` itself also keeps storing them, so my attempt to forcibly update tracker entries when the torrent is stopped was unsuccessful. All I can do is reset tracker entries locally (in qBittorrent code). This looks quite consistent to me, since they look the same as if they were stopped when starting qBittorrent.

glassez commented:
> Shouldn't `Hide zero and infinity values` apply to these new columns?

Why not?
But how about other columns? As I can see this option isn't used in tracker list currently so I would prefer if it is considered as a separate issue.

thalieht commented:
>Why not?

I meant it should apply. Zero for other columns has meaning (no seeds peers etc) but in these new columns it's just spam IMO.

glassez commented:
> I meant it should apply. Zero for other columns has meaning (no seeds peers etc) but in these new columns it's just spam IMO.

I suspect infinity signs are being shown in an incorrect context. IIRC, infinity Next/Min announce is displayed when the corresponding time point goes into the past (i.e. it becomes earlier than the current time). Shouldn't it be displayed as `Exp.` (i.e. Expired) in this case?

thalieht commented:
>when the corresponding time point goes into the past (i.e. it becomes earlier than the current time).

I'm afraid i don't know even know if that's normal behavior so i'm not one to ask anything related to it.

glassez commented:
> I'm afraid i don't know even know if that's normal behavior

This may be in the quite usual case when the timeout has just expired, but it has not yet had time to send an announce and, accordingly, schedule the next one.

glassez commented:
> Also the peers and times downloaded column for DHT, PEX and LSD seems to remain empty and it looks odd. Not sure if that was intended. ![image](https://user-images.githubusercontent.com/141441441/268074709-19e5e54b-8af4-416b-aea6-4a6ac1c3e822.png)

Is it regression of this PR?

glassez commented:
> At present, it appears that the commit erases all relevant information and marks it as "not contacted yet."

So what should be displayed instead? Should I just clear (or set to 0) announce times? And maybe other statistics (peers, seeds, etc.)?

stalkerok commented:
IMO, everything works as it should, the only thing I would remove `yet`

glassez commented:
> IMO, everything works as it should, the only thing I would remove `yet`

Do you mean you're OK with resetting tracker entries once torrent is paused as it is done currently? Except "Not contacted yet" should be changed to "Not contacted"?

stalkerok commented:
> Do you mean you're OK with resetting tracker entries once torrent is paused as it is done currently? Except "Not contacted yet" should be changed to "Not contacted"?

Yes, since there was a contact before, it would be better to change it to "Not contacted" or "No connection"

glassez commented:
`Working`, `Worked`... IMO, it's too complicated... they both mean that the last announce was successful. So maybe just replace it with `OK`?

stalkerok commented:
> But we do not get to know if that contact was successful or not.

For a stopped torrent, this is not important. To get fresh data, just run it.

> `Working`, `Worked`... IMO, it's too complicated... they both mean that the last announce was successful. So maybe just replace it with `OK`?

The status of "Working" is definitely better not to change. What about the "Torrent stopped" status?

glassez commented:
> "Torrent stopped"

I don't like to use "Torrent stopped" as tracker-related status. Moreover it is obvious when you see the trackers of stopped torrent.

The question is, what is the status supposed to mean? If this is the status of the last announce, then why shouldn't it be exactly this? i.e. display the status of the last announce, even if the torrent was stopped afterwards?

stalkerok commented:
> If this is the status of the last announce, then why shouldn't it be exactly this? i.e. display the status of the last announce, even if the torrent was stopped afterwards?

This will probably be best.

glassez commented:
> > If this is the status of the last announce, then why shouldn't it be exactly this? i.e. display the status of the last announce, even if the torrent was stopped afterwards?
> 
> This will probably be best.

I tried to implement it in #19633.
